File: cpp-server/src/roomManager.cpp
#include "roomManager.h"
#include "session.h"
#include "server.h"
#include "TwitchClient.h"      // fixes TwitchClient errors
#include <iostream>

using json = nlohmann::json;

void RoomManager::joinRoom(const std::string& roomId, std::shared_ptr<Session> s, const std::string& username) {
    std::lock_guard<std::mutex> lock(m_mutex);
    m_rooms[roomId].join(s, username);
}

void RoomManager::leaveAll(std::shared_ptr<Session> s) {
    std::lock_guard<std::mutex> lock(m_mutex);
    for (auto& pair : m_rooms) {
        const std::string& id = pair.first;
        Room& room = pair.second;

        if (room.leave(s)) {
            json sysMsg = {
                {"type", "system"},
                {"room", id},
                {"payload", "Streamer disconnected, lobby cleared"}
            };
            room.broadcast(sysMsg.dump());

            // clear players too if streamer disconnects
            room.resetLobby();
        }
    }
}

static std::string normalizeRoom(const std::string& roomId) {
    if (!roomId.empty() && roomId[0] == '#')
        return roomId.substr(1);
    return roomId;
}
void RoomManager::handleJoin(std::shared_ptr<Session> s, const nlohmann::json& j, const std::string& roomId) {
    // Clean up any abandoned rooms first
    cleanupAbandonedRooms();
    
    // Also clean up expired rooms (inactive for 1+ hours)
    cleanupExpiredRooms();

    std::string username;

    // Accept both string and object payloads
    if (j["payload"].is_string()) {
        username = j["payload"].get<std::string>();
    }
    else if (j["payload"].is_object()) {
        username = j["payload"].value("username", "");
    }

    if (username.empty()) return;

    Room* room;
    bool isNewRoom = false;
    {
        std::lock_guard<std::mutex> lock(m_mutex);
        auto it = m_rooms.find(roomId);
        if (it == m_rooms.end()) {
            // This is a new room being created
            isNewRoom = true;
            std::cout << "[ROOM] Creating new room: " << roomId << std::endl;
        }
        room = &m_rooms[roomId];
    }
    
    // If this is a new room, set it as the current room for the Twitch bot
    if (isNewRoom && m_server) {
        // Extract channel from the join message
        std::string channel = j.value("channel", "");
        if (!channel.empty()) {
            // Store the channel this room belongs to
            m_roomChannels[roomId] = channel;
            std::cout << "[ROOM] Room " << roomId << " belongs to channel " << channel << std::endl;
            
            // Set this as the current room for that channel's Twitch bot
            m_server->setCurrentRoom("#" + channel, roomId);
        } else {
            std::cout << "[WARN] No channel specified for new room " << roomId << std::endl;
        }
    }

    if (room->hasPlayer(username)) {
        std::cout << "[SPAM] Duplicate join from " << username << " (replaying state)\n";
        if (s) {
            room->join(s, username);     // attach new session
            room->replayPlayers(s);      // send full player list
            room->replayHistory(s);      // send all strokes
        }
        return;
    }

    // First-time join
    room->join(s, username);
}




void RoomManager::handleLeave(std::shared_ptr<Session> s, const nlohmann::json& j, const std::string& roomId) {
    auto it = m_rooms.find(roomId);
    if (it != m_rooms.end()) {
        Room& room = it->second;

        for (auto& [uname, p] : room.getPlayers()) {
            if (room.leave(s)) {
                nlohmann::json leaveMsg = {
                    {"type", "leave"},
                    {"payload", {
                        {"id", p.id},
                        {"username", uname}
                    }}
                };
                room.broadcast(leaveMsg.dump());
            }
        }

        // Clean up abandoned rooms
        if (room.empty()) {
            std::cout << "[ROOM] Room " << roomId << " is empty, removing it" << std::endl;
            m_rooms.erase(it);
        }
    }
}


void RoomManager::handleChat(std::shared_ptr<Session>, const json& j, const std::string& roomId) {
    std::string payload = j.value("payload", "");
    if (!roomId.empty() && !payload.empty()) {
        json chatMsg = { {"type","chat"}, {"room",roomId}, {"payload",payload} };
        m_rooms[roomId].broadcast(chatMsg.dump());
    }
}

void RoomManager::handleEndRound(const std::string& roomId) {
    auto it = m_rooms.find(roomId);
    if (it != m_rooms.end()) {
        it->second.endRound();
    }
}

void RoomManager::handleStopBot(const json& j) {
    std::string channel = j.value("channel", "");
    if (m_server) {
        m_server->stopBot(channel);
        std::cout << "ADMIN Stopped Twitch bot for channel: " << channel << "\n";
    }
}

void RoomManager::handleSpawnBot(const json& j) {
    std::string oauth = j.value("oauth", "");
    std::string nick = j.value("nick", "");
    std::string channel = j.value("channel", "");

    if (m_server) {
        bool spawned = m_server->spawnBot(oauth, nick, channel);
        if (spawned) {
            std::cout << "ADMIN Spawned Twitch bot for channel: " << channel << "\n";
        }
        else {
            std::cout << "[ADMIN] Bot for channel " << channel << " already exists, ignoring spawn.\n";
        }
    }
}

void RoomManager::handleStatus(const std::string& jsonMsg) {
    if (m_server) {
        m_server->broadcast(jsonMsg);
    }
}

void RoomManager::handleDraw(std::shared_ptr<Session> s, const json& j, const std::string& roomId) {
    if (roomId.empty()) return;

    json drawMsg = {
        {"type", "draw"},
        {"room", roomId},
        {"payload", j["payload"]}
    };

    // store in room history
    m_rooms[roomId].addStroke(drawMsg);

    // broadcast to all
    m_rooms[roomId].broadcast(drawMsg.dump());
}

void RoomManager::handleClear(std::shared_ptr<Session> s, const json& j, const std::string& roomId) {
    if (roomId.empty()) return;

    json clearMsg = {
        {"type", "clear"},
        {"room", roomId}
    };

    // clear room history
    m_rooms[roomId].clearHistory();

    // broadcast clear
    m_rooms[roomId].broadcast(clearMsg.dump());
}

void RoomManager::handleRestoreState(std::shared_ptr<Session> s, const std::string& roomId) {
    std::cout << "[DEBUG] handleRestoreState called for room: " << roomId << std::endl;
    if (roomId.empty() || !s) return;

    auto it = m_rooms.find(roomId);
    if (it != m_rooms.end()) {
        const Room& room = it->second;

        // Get data outside of any potential locks
        std::vector<std::string> playerUsernames;
        std::vector<json> strokeHistory;

        {
            std::lock_guard<std::mutex> lock(m_mutex);
            // Fix: Get the data first, then copy it properly
            auto usernames = room.getPlayerUsernames();
            playerUsernames.assign(usernames.begin(), usernames.end());
            strokeHistory = room.getStrokeHistory();
        }

        // Send current state back to client
        json response;
        response["type"] = "current_state";
        response["payload"]["players"] = playerUsernames;
        response["payload"]["strokes"] = strokeHistory;

        std::cout << "[DEBUG] About to send state with " << strokeHistory.size() << " strokes" << std::endl;
        s->send(response.dump());
        std::cout << "[STATE] Sent current state to client for room: " << roomId << std::endl;
    }
    else {
        std::cout << "[DEBUG] Room not found: " << roomId << std::endl;
    }
}

void RoomManager::cleanupAbandonedRooms() {
    std::lock_guard<std::mutex> lock(m_mutex);

    auto it = m_rooms.begin();
    while (it != m_rooms.end()) {
        if (it->second.empty()) {
            std::cout << "[ROOM] Cleaning up abandoned room: " << it->first << std::endl;
            it = m_rooms.erase(it);
        }
        else {
            ++it;
        }
    }
}

void RoomManager::cleanupExpiredRooms() {
    std::lock_guard<std::mutex> lock(m_mutex);
    
    auto now = std::chrono::steady_clock::now();
    auto oneHour = std::chrono::hours(1);
    
    auto it = m_rooms.begin();
    while (it != m_rooms.end()) {
        auto lastActivity = it->second.getLastActivity();
        if (now - lastActivity > oneHour) {
            std::cout << "[ROOM] Cleaning up expired room: " << it->first 
                      << " (inactive for " << std::chrono::duration_cast<std::chrono::minutes>(now - lastActivity).count() << " minutes)" << std::endl;
            
            // Remove from room channels tracking
            m_roomChannels.erase(it->first);
            
            it = m_rooms.erase(it);
        } else {
            ++it;
        }
    }
}


void RoomManager::onMessage(std::shared_ptr<Session> s, const std::string& jsonMsg) {
    try {
        auto j = json::parse(jsonMsg);
        std::string type = j.value("type", "");
        std::string roomId = normalizeRoom(j.value("room", ""));

        if (type == "join")        handleJoin(s, j, roomId);
        else if (type == "leave")  handleLeave(s, j, roomId);
        else if (type == "chat")   handleChat(s, j, roomId);
        else if (type == "end_round") handleEndRound(roomId);
        else if (type == "stop_bot")  handleStopBot(j);
        else if (type == "spawn_bot") handleSpawnBot(j);
        else if (type == "status")    handleStatus(jsonMsg);
        else if (type == "pong" && s) s->markPongReceived();
        else if (type == "draw")      handleDraw(s, j, roomId);
        else if (type == "clear")     handleClear(s, j, roomId);
        else if (type == "get_state") handleRestoreState(s, roomId);
        else {
            std::cerr << "[WARN] Unknown type: " << type << " msg=" << jsonMsg << "\n";
        }
    }
    catch (const std::exception& e) {
        std::cerr << "[ERROR] onMessage parse failed: " << e.what()
            << " raw=" << jsonMsg << "\n";
    }
}

File: game-server/src/roomManager.ts
import type { ClientSession, GameMessage } from "./types.js";
import { Room } from "./room.js";
import type { GameServer } from "./gameServer.js";

function normalizeRoom(roomId: string): string {
  if (roomId.startsWith("#")) return roomId.slice(1);
  return roomId;
}

export class RoomManager {
  private rooms = new Map<string, Room>();
  /** roomId -> twitch channel name (no #) */
  private roomChannels = new Map<string, string>();
  private server: GameServer | null = null;

  setServer(server: GameServer): void {
    this.server = server;
  }

  getCurrentRoom(channel: string): Room | undefined {
    const normalized = normalizeRoom(channel);
    for (const [roomId, ch] of this.roomChannels) {
      if (ch.toLowerCase() === normalized) {
        return this.rooms.get(roomId);
      }
    }
    return undefined;
  }

  leaveAll(session: ClientSession): void {
    for (const room of this.rooms.values()) {
      room.leave(session);
    }
  }

  onMessage(session: ClientSession | null, raw: string): void {
    try {
      const j = JSON.parse(raw) as GameMessage;
      const type = j.type ?? "";
      const roomId = normalizeRoom(j.room ?? "");

      switch (type) {
        case "join":
          this.handleJoin(session, j, roomId);
          break;
        case "leave":
          this.handleLeave(session, j, roomId);
          break;
        case "chat":
          this.handleChat(j, roomId);
          break;
        case "start_round":
          this.handleStartRound(roomId, j);
          break;
        case "guess":
          this.handleGuessBlocked();
          break;
        case "end_round":
          this.handleEndRound(roomId);
          break;
        case "stop_bot":
          this.handleStopBot(session, j);
          break;
        case "spawn_bot":
          this.handleSpawnBot(session, j);
          break;
        case "map_twitch_room":
          this.handleMapTwitchRoom(j);
          break;
        case "status":
          if (this.server) this.server.broadcast(raw);
          break;
        case "pong":
          session?.markPongReceived();
          break;
        case "draw":
          this.handleDraw(roomId, j);
          break;
        case "clear":
          this.handleClear(roomId);
          break;
        case "undo":
          this.handleUndo(roomId);
          break;
        case "get_state":
          if (session) this.handleRestoreState(session, roomId);
          break;
        default:
          console.warn(`[WARN] Unknown type: ${type}`, raw);
      }
    } catch (e) {
      console.error("[ERROR] onMessage parse failed:", e, raw);
    }
  }

  private getOrCreateRoom(roomId: string): Room {
    let room = this.rooms.get(roomId);
    if (!room) {
      room = new Room();
      this.wireRoomChatAnnouncements(room, roomId);
      this.rooms.set(roomId, room);
      console.log(`[ROOM] Creating new room: ${roomId}`);
    }
    return room;
  }

  private wireRoomChatAnnouncements(room: Room, roomId: string): void {
    room.setChatAnnouncements({
      onRoundStart: () => {
        this.announceToTwitch(
          roomId,
          "Round starting — type !guess <word> in chat now!",
        );
      },
      onCorrectGuess: (username) => {
        this.announceToTwitch(roomId, `${username} got it right!`);
      },
      onRoundTimeout: (word) => {
        this.announceToTwitch(roomId, `Time's up! The word was: ${word}`);
      },
      onRoundSkipped: (word) => {
        this.announceToTwitch(roomId, `Round skipped. The word was: ${word}`);
      },
    });
  }

  private announceToTwitch(roomId: string, message: string): void {
    const channel = this.roomChannels.get(roomId);
    if (!channel || !this.server) return;
    this.server.sayInChannel(channel, message);
  }

  private handleJoin(
    session: ClientSession | null,
    j: GameMessage,
    roomId: string,
  ): void {
    this.cleanupAbandonedRooms();
    this.cleanupExpiredRooms();

    let username = "";
    if (typeof j.payload === "string") {
      username = j.payload;
    } else if (j.payload && typeof j.payload === "object" && "username" in j.payload) {
      username = String((j.payload as { username?: string }).username ?? "");
    }
    if (!username || !roomId) return;

    const isNewRoom = !this.rooms.has(roomId);
    const room = this.getOrCreateRoom(roomId);
    const channel = (j.channel ?? "").replace(/^#/, "").toLowerCase();

    if (!isNewRoom && channel && this.server) {
      this.roomChannels.set(roomId, channel);
      this.server.setCurrentRoom(channel, roomId);
      console.log(`[ROOM] Reconnecting to existing room: ${roomId}`);
    }

    if (isNewRoom && channel && this.server) {
      for (const [rid, ch] of [...this.roomChannels]) {
        if (ch === channel) this.roomChannels.delete(rid);
      }
      this.roomChannels.set(roomId, channel);
      this.server.setCurrentRoom(channel, roomId);
      console.log(`[ROOM] Bot connected to new room: ${roomId}`);
    }

    if (room.hasPlayer(username)) {
      console.log(`[SPAM] Duplicate join from ${username}`);
      if (session) room.join(session, username);
      return;
    }

    room.join(session, username);
  }

  private handleLeave(
    session: ClientSession | null,
    j: GameMessage,
    roomId: string,
  ): void {
    const room = this.rooms.get(roomId);
    if (!room || !session) return;

    if (j.intentional) {
      for (const [uname, p] of room.getPlayers()) {
        room.broadcast(
          JSON.stringify({
            type: "leave",
            payload: { id: p.id, username: uname },
          }),
        );
      }
      room.resetLobby();
      room.clearHistory();
    }

    room.leave(session);
    if (room.isEmpty()) {
      console.log(`[ROOM] Room ${roomId} is empty, removing`);
      this.rooms.delete(roomId);
      this.roomChannels.delete(roomId);
    }
  }

  private handleChat(j: GameMessage, roomId: string): void {
    const payload = typeof j.payload === "string" ? j.payload : "";
    const room = this.rooms.get(roomId);
    if (!room || !payload) return;
    room.broadcast(
      JSON.stringify({ type: "chat", room: roomId, payload }),
    );
  }

  private handleStartRound(roomId: string, j: GameMessage): void {
    const room = this.rooms.get(roomId);
    if (!room) return;
    let word = "apple";
    if (j.payload && typeof j.payload === "object" && "word" in j.payload) {
      word = String((j.payload as { word?: string }).word ?? word);
    }
    room.startRound(word);
  }

  private handleGuessBlocked(): void {
    console.log("[SECURITY] Blocked direct guess attempt from WebSocket");
  }

  private handleEndRound(roomId: string): void {
    const room = this.rooms.get(roomId);
    if (room) room.endRound();
  }

  private handleStopBot(session: ClientSession | null, j: GameMessage): void {
    const channel = j.channel ?? "";
    if (!this.server) return;

    const room = this.getCurrentRoom(channel);
    if (room) {
      console.log(`[ROOM] Clearing lobby before stop_bot for ${channel}`);
      room.resetLobby();
    }

    this.server.stopBot(channel);
    console.log(`ADMIN Stopped Twitch bot for channel: ${channel}`);

    if (session) {
      session.send(
        JSON.stringify({
          type: "status",
          status: "ok",
          message: "Bot stopped",
          channel,
        }),
      );
    }
  }

  private handleSpawnBot(session: ClientSession | null, j: GameMessage): void {
    const oauth = j.oauth ?? "";
    const nick = j.nick ?? "";
    const channel = j.channel ?? "";
    const roomId = normalizeRoom(j.room_id ?? j.room ?? "");
    if (!this.server) return;

    const spawned = this.server.spawnBot(oauth, nick, channel);
    const message = spawned
      ? `Spawned Twitch bot for ${channel}`
      : `Bot for ${channel} already exists`;

    if (spawned) console.log(`ADMIN ${message}`);
    else console.log(`[ADMIN] ${message}`);

    if (roomId && channel) {
      const channelName = channel.replace(/^#/, "").toLowerCase();
      for (const [rid, ch] of [...this.roomChannels]) {
        if (ch.toLowerCase() === channelName) this.roomChannels.delete(rid);
      }
      this.getOrCreateRoom(roomId);
      this.roomChannels.set(roomId, channelName);
      this.server.setCurrentRoom(channelName, roomId);
      console.log(`[ROOM] Mapped ${channelName} -> ${roomId} via spawn_bot`);
    }

    if (session) {
      session.send(
        JSON.stringify({
          type: "status",
          status: spawned ? "ok" : "exists",
          message,
          channel,
        }),
      );
    }
  }

  private handleMapTwitchRoom(j: GameMessage): void {
    const payload = (j.payload ?? {}) as { twitch_name?: string; room_id?: string };
    const twitchName = payload.twitch_name ?? "";
    const roomId = payload.room_id ?? "";
    if (!twitchName || !roomId) return;

    this.roomChannels.set(roomId, twitchName);
    if (this.server) {
      this.server.setCurrentRoom(twitchName, roomId);
    }
  }

  private handleDraw(roomId: string, j: GameMessage): void {
    const room = this.rooms.get(roomId);
    if (!room) return;

    const drawMsg = { type: "draw", room: roomId, payload: j.payload };
    room.addStroke(drawMsg);
    room.broadcast(JSON.stringify(drawMsg));
  }

  private handleClear(roomId: string): void {
    const room = this.rooms.get(roomId);
    if (!room) return;
    room.clearHistory();
    room.broadcast(JSON.stringify({ type: "clear", room: roomId }));
  }

  private handleUndo(roomId: string): void {
    const room = this.rooms.get(roomId);
    if (!room) return;
    if (!room.undoLastStroke()) return;
    room.broadcast(
      JSON.stringify({
        type: "undo",
        room: roomId,
        payload: { strokes: room.getStrokeHistory() },
      }),
    );
  }

  private handleRestoreState(session: ClientSession, roomId: string): void {
    const room = this.rooms.get(roomId);
    if (!room) {
      console.log(`[DEBUG] Room not found: ${roomId}`);
      return;
    }

    const round = room.getCurrentRound();
    const payload: Record<string, unknown> = {
      players: room.getPlayerUsernames(),
      strokes: room.getStrokeHistory(),
    };

    if (round.active) {
      const elapsed = Math.floor((Date.now() - round.startTime) / 1000);
      let timeLeft = round.duration - elapsed;
      if (timeLeft < 0) timeLeft = 0;
      payload.round = {
        active: true,
        word: round.word,
        hint: round.hint,
        timeLeft,
      };
    }

    session.send(
      JSON.stringify({ type: "current_state", payload }),
    );
    console.log(`[STATE] Sent current state for room: ${roomId}`);
  }

  private cleanupAbandonedRooms(): void {
    for (const [id, room] of this.rooms) {
      if (room.isEmpty()) {
        console.log(`[ROOM] Cleaning up abandoned room: ${id}`);
        this.rooms.delete(id);
        this.roomChannels.delete(id);
      }
    }
  }

  private cleanupExpiredRooms(): void {
    const oneHour = 60 * 60 * 1000;
    const now = Date.now();
    for (const [id, room] of this.rooms) {
      if (now - room.getLastActivity() > oneHour) {
        console.log(`[ROOM] Cleaning up expired room: ${id}`);
        this.rooms.delete(id);
        this.roomChannels.delete(id);
      }
    }
  }
}
